### PR TITLE
Verified monads

### PR DESCRIPTION
**Pull request by marcosh**
@clayrat could you please take a quick look? especially at the postulates I'm introducting

### CODE REVIEW

File: src/Monad/IOMonad.lidr [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+> ioMLeftIdentity :
+>      (x : a)
+>   -> (f : a -> IO' ffi b)
+>   -> pure x >>= f = f x
+>
+> ioAMap :
+>      (io : IO' ffi a)
+>   -> (f : a -> b)
+>   -> map f io = pure f <*> io
+> ioAMap io f = rewrite ioMLeftIdentity f (\g => io >>= (pure . g)) in Refl
```
Comment (clayrat) [inline src/Monad/IOMonad.lidr]:
```suggestion
> ioAMap io f = sym $ ioMLeftIdentity f (\g => io >>= (pure . g))
```

File: src/Monad/IOMonad.lidr [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+>   applicativeMap = ioAMap
+>   applicativeIdentity = ioAIdentity
+>   applicativeComposition = ioAComposition
+>   applicativeHomomorphism = ioAHomomorphism
+>   applicativeInterchange = ioAInterchange
+>
+> ioMApplicative :
+>      (mf : IO' ffi (a -> b))
+>   -> (io : IO' ffi a)
+>   -> mf <*> io = mf >>= \f =>
```
Comment (clayrat) [inline src/Monad/IOMonad.lidr]: It's also possible to use do-notation:
```suggestion
>   -> mf <*> io = do f <- mf
>                     x <- io    
>                     pure (f x) 
```

File: src/Monad/Monad.lidr [hunk 3]
```diff
@@ -84,6 +84,48 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
...
+>                                    (functorComposition cat cat cat functor functor)
+>                                    functor
+>                                    (replace {P = \x => NaturalTransformation cat cat x (functorComposition cat cat cat functor functor)}
+>                                             (catsLeftIdentity cat cat functor)
+>                                             (composeFunctorNatTrans cat cat cat
+>                                                                     (idFunctor cat)
+>                                                                     functor
+>                                                                     unit
+>                                                                     functor))
+>                                             multiplication
```
Comment (clayrat) [inline src/Monad/Monad.lidr]: alignment

File: src/Monad/VerifiedMonadAsMonad.lidr [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+> import Syntax.PreorderReasoning
+>
+> -- contrib
+> import Interfaces.Verified
+>
+> %access public export
+> %default total
+>
+> verifiedMonadToFunctor : VerifiedMonad m => Functor m
+> verifiedMonadToFunctor @{monad} = %implementation
```
Comment (clayrat) [inline src/Monad/VerifiedMonadAsMonad.lidr]: Actually none of these `@{monad}` constraints are required

File: src/Monad/VerifiedMonadAsMonad.lidr [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+>
+> verifiedMonadToVerifiedApplicative : VerifiedMonad m => VerifiedApplicative m
+> verifiedMonadToVerifiedApplicative @{monad} = %implementation
+>
+> verifiedMonadToMonad : VerifiedMonad m => Monad m
+> verifiedMonadToMonad @{monad} = %implementation
+>
+> verifiedMonadToCFunctor : VerifiedMonad m => CFunctor TypesAsCategoryExtensional.typesAsCategoryExtensional
+>                                                       TypesAsCategoryExtensional.typesAsCategoryExtensional
+> verifiedMonadToCFunctor @{monad} = functorToCFunctor $ verifiedMonadToVerifiedFunctor @{monad}
```
Comment (clayrat) [inline src/Monad/VerifiedMonadAsMonad.lidr]:
```suggestion
> verifiedMonadToCFunctor {m} = functorToCFunctor $ verifiedMonadToVerifiedFunctor {m}
```

**Comment by clayrat** [comment]
@marcosh I rewrote `VerifiedMonadAsMonad` in a more idiomatic way and pushed it directly to this branch, seemed like a faster way, hope you don't mind?

**Comment by marcosh** [comment]
@clayrat not at all, thanks for it!

**Pull request merged**